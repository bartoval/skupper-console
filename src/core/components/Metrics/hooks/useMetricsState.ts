import { useCallback, useState, startTransition } from 'react';

import { useMetricSessionHandlers } from './useMetricsSessionHandler';
import { QueryMetricsParams } from '../../../../types/Metrics.interfaces';

interface UseMetricsProps {
  sessionKey?: string;
  defaultMetricFilterValues: QueryMetricsParams;
}

export const useMetricsState = ({ sessionKey, defaultMetricFilterValues }: UseMetricsProps) => {
  const { selectedFilters: storedFilters, setSelectedFilters: setStoredFilters } = useMetricSessionHandlers(
    sessionKey,
    defaultMetricFilterValues
  );

  const [selectedFilters, setSelectedFilters] = useState<QueryMetricsParams>(storedFilters);

  const handleFilterChange = useCallback(
    (updatedFilters: QueryMetricsParams) => {
      startTransition(() => {
        // let to override undefined values with default values if they exists
        const filters = { ...defaultMetricFilterValues, ...JSON.parse(JSON.stringify({ ...updatedFilters })) };

        setStoredFilters(filters);
        setSelectedFilters(filters);
      });
    },
    [defaultMetricFilterValues, setStoredFilters]
  );

  return {
    selectedFilters,
    handleFilterChange
  };
};
