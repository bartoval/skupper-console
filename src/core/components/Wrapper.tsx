import { ReactElement } from 'react';

import { QueryClient, QueryClientConfig, QueryClientProvider } from '@tanstack/react-query';
import { HashRouter } from 'react-router-dom';

import { queryClientConfig } from '../../config/reactQuery';

const QueryClientContext = function ({
  config = {},
  children
}: {
  config?: QueryClientConfig;
  children: ReactElement;
}) {
  const queryClient = new QueryClient({
    ...queryClientConfig,
    ...config
  });

  return <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>;
};

export const Wrapper = function ({ children, config }: { config?: QueryClientConfig; children: ReactElement }) {
  return (
    <HashRouter
      future={{
        v7_relativeSplatPath: true,
        v7_startTransition: false
      }}
    >
      <QueryClientContext config={config}>{children}</QueryClientContext>
    </HashRouter>
  );
};
