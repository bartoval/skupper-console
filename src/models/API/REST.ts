import { MonitoringRoutersTopology } from '@pages/Monitoring/services/services.interfaces';

import { fetchWithTimeout } from './Middleware';
import {
  DATA_URL,
  MONITORING_FLOWS,
  LINKS,
  SITE_URL,
  SITES_SERVICES,
  TARGETS,
  MONITORING_ROUTERS_STAT,
  MONITORING_STATS,
  MONITORING_SERVICES_STATS,
  MONITORING_ROUTERS_TOPOLOGY,
  MONITORING_CONNECTIONS,
  SERVICES,
} from './REST.constant';
import {
  DataResponse,
  FlowsResponse,
  LinkResponse,
  MonitoringStatsResponse,
  RouterStatsResponse,
  ServiceSitesResponse,
  TargetResponse,
  ServicesStatsResponse,
  ServiceResponse,
} from './REST.interfaces';

export const RESTApi = {
  fetchData: async (): Promise<DataResponse> => {
    const { data } = await fetchWithTimeout(DATA_URL);

    return data;
  },
  fetchSite: async (): Promise<string> => {
    const { data } = await fetchWithTimeout(SITE_URL);

    return data;
  },
  fetchLinks: async (): Promise<LinkResponse[]> => {
    const { data } = await fetchWithTimeout(LINKS);

    return data;
  },
  fetchTargets: async (): Promise<TargetResponse[]> => {
    const { data } = await fetchWithTimeout(TARGETS);

    return data;
  },
  fetchSitesServices: async (): Promise<ServiceSitesResponse[]> => {
    const { data } = await fetchWithTimeout(SITES_SERVICES);

    return data;
  },
  fetchFlows: async (): Promise<FlowsResponse[]> => {
    const { data } = await fetchWithTimeout(MONITORING_FLOWS);

    return data;
  },
  fetchMonitoringConnectionsByVanId: async (vanaddr: string): Promise<FlowsResponse[]> => {
    const { data } = await fetchWithTimeout(MONITORING_CONNECTIONS, { params: { vanaddr } });

    return data;
  },
  fetchMonitoringFlowsByVanId: async (vanaddr: string): Promise<FlowsResponse[]> => {
    const { data } = await fetchWithTimeout(MONITORING_FLOWS, { params: { vanaddr } });

    return data;
  },
  fetchVANAddresses: async (): Promise<FlowsResponse[]> => {
    const { data } = await fetchWithTimeout(MONITORING_FLOWS);

    return data;
  },

  // SERVICES APIs
  fetchServices: async (): Promise<ServiceResponse[]> => {
    const { data } = await fetchWithTimeout(SERVICES);

    return data;
  },

  // FLOWS APIs
  fetchFlowsStats: async (): Promise<MonitoringStatsResponse[]> => {
    const { data } = await fetchWithTimeout(MONITORING_STATS);

    return data;
  },
  fetchFlowsRoutersStats: async (): Promise<RouterStatsResponse[]> => {
    const { data } = await fetchWithTimeout(MONITORING_ROUTERS_STAT);

    return data;
  },
  fetchFlowsServicesStats: async (): Promise<ServicesStatsResponse[]> => {
    const { data } = await fetchWithTimeout(MONITORING_SERVICES_STATS);

    return data;
  },
  fetchTopologyRoutersLinks: async (): Promise<MonitoringRoutersTopology> => {
    const { data } = await fetchWithTimeout(MONITORING_ROUTERS_TOPOLOGY);

    return data;
  },
};
