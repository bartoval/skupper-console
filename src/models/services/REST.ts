import { RESTApi } from '../API/REST';
import { ServicesResponse, TargetsResponse } from '../API/REST.interfaces';
import Adapter from './adapter';
import { Data, DataVAN, SiteInfoService } from './REST.interfaces';

export const RESTServices = {
  fetchData: async (): Promise<Data> => {
    const [dataVAN, siteId, links, targets, services] = await Promise.all([
      RESTApi.fetchData(),
      RESTApi.fetchSite(),
      RESTApi.fetchLinks(),
      RESTApi.fetchTargets(),
      RESTApi.fetchServices(),
    ]);
    const data: DataVAN = new Adapter(dataVAN).getData();

    const { site_id, site_name, namespace } =
      data.sites.find(({ site_id: id }) => id === siteId) || data.sites[0];

    return {
      data,
      siteInfo: {
        links,
        targets,
        services: normalizeServices(targets, services),
        siteId: site_id,
        siteName: site_name,
        namespace,
      },
    };
  },
};

function normalizeServices(
  targets: TargetsResponse[],
  services: ServicesResponse[],
): SiteInfoService[] {
  const servicesNormalized: SiteInfoService[] = [];

  services.forEach((service) => {
    const serviceTarget = targets.find(({ name }) => name === service.name);

    if (!serviceTarget) {
      servicesNormalized.push(service);

      return;
    }

    targets.forEach(({ name, type, ports }) => {
      const endpoint = service?.endpoints?.find(({ target }) => target === name);

      if (service && endpoint) {
        servicesNormalized.push({ ...service, type, exposed: true, ports });
      } else {
        servicesNormalized.push({ type, name, exposed: false });
      }
    });
  });

  return servicesNormalized;
}
