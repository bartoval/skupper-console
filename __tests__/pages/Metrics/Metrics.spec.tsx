import { Suspense } from 'react';

import { act, render, renderHook } from '@testing-library/react';

import { useMetricsState } from '../../../src/pages/shared/Metrics/hooks/useMetricsState';

import { Protocols } from '../../../src/API/REST.enum';
import { Wrapper } from '../../../src/core/components/Wrapper';
import LoadingPage from '../../../src/pages/shared/Loading';
import Metrics, { MetricsProps } from '../../../src/pages/shared/Metrics';
import { configDefaultFilters } from '../../../src/pages/shared/Metrics/Metrics.constants';

describe('useMetrics', () => {
  const initialProps = {
    defaultMetricFilterValues: { protocol: Protocols.Http },
    defaultOpenSections: { byterate: true },
    onGetMetricFiltersConfig: jest.fn(),
    onGetExpandedSectionsConfig: jest.fn()
  };

  it('initializes with the correct state from props', () => {
    const { result } = renderHook(() => useMetricsState(initialProps));
    expect(result.current.queryParams).toEqual({ protocol: Protocols.Http });
    expect(result.current.shouldUpdateData).toBe(0);
  });

  it('updates shouldUpdateData correctly when triggerMetricUpdate is called', () => {
    const { result } = renderHook(() => useMetricsState(initialProps));
    act(() => {
      result.current.triggerMetricUpdate();
    });
    expect(result.current.shouldUpdateData).not.toBe(0);
  });

  it('updates queryParams and calls onGetMetricFiltersConfig when handleFilterChange is invoked', () => {
    const { result } = renderHook(() => useMetricsState(initialProps));
    const newFilters = { protocol: Protocols.Tcp };
    act(() => {
      result.current.handleFilterChange(newFilters);
    });
    expect(result.current.queryParams).toEqual(newFilters);
    expect(initialProps.onGetMetricFiltersConfig).toHaveBeenCalledWith(newFilters);
  });

  it('updates expanded sections and calls onGetExpandedSectionsConfig when handleSectionToggle is invoked', () => {
    const { result } = renderHook(() => useMetricsState(initialProps));
    const newSections = { traffic: false };
    act(() => {
      result.current.handleSectionToggle(newSections);
    });
    expect(initialProps.onGetExpandedSectionsConfig).toHaveBeenCalledWith({
      ...initialProps.defaultOpenSections,
      ...newSections
    });
  });
});

describe('Metrics Component', () => {
  const setup = (overrides?: Partial<MetricsProps>) => ({
    configFilters: configDefaultFilters,
    defaultMetricFilterValues: {},
    defaultOpenSections: {},
    sourceSites: [],
    destSites: [],
    sourceProcesses: [],
    destProcesses: [],
    availableProtocols: [Protocols.Http, Protocols.Tcp, Protocols.Http2],
    onGetMetricFiltersConfig: jest.fn(),
    onGetExpandedSectionsConfig: jest.fn(),
    ...overrides
  });

  it('renders correctly', () => {
    const { asFragment } = render(
      <Wrapper>
        <Suspense fallback={<LoadingPage />}>
          <Metrics {...setup()} />
        </Suspense>
      </Wrapper>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
