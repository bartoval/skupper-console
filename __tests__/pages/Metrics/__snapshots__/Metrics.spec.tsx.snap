// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metrics Component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="pf-v5-l-stack pf-m-gutter"
  >
    <div
      class="pf-v5-l-stack__item"
      style="position: sticky; top: 0px; z-index: 1;"
    >
      <section
        class="pf-v5-c-page__main-section pf-m-no-padding"
      >
        <div
          class="pf-v5-c-toolbar"
          data-ouia-component-id="OUIA-Generated-Toolbar-1"
          data-ouia-component-type="PF5/Toolbar"
          data-ouia-safe="true"
          id="pf-random-id-0"
        >
          <div
            class="pf-v5-c-toolbar__content"
          >
            <div
              class="pf-v5-c-toolbar__content-section"
            >
              <div
                class="pf-v5-c-toolbar__group"
              >
                <div
                  class="pf-v5-c-toolbar__item"
                >
                  <button
                    aria-expanded="false"
                    class="pf-v5-c-menu-toggle"
                    data-ouia-component-id="OUIA-Generated-MenuToggle-1"
                    data-ouia-component-type="PF5/MenuToggle"
                    data-ouia-safe="true"
                    type="button"
                  >
                    <span
                      class="pf-v5-c-menu-toggle__icon"
                    >
                      <div
                        style="display: contents;"
                      >
                        <span
                          class="sk-resource-icon sk-resource-site"
                          role="site-resource-icon"
                        >
                          S
                        </span>
                      </div>
                    </span>
                    <span
                      class="pf-v5-c-menu-toggle__text"
                    >
                      All source sites
                    </span>
                    <span
                      class="pf-v5-c-menu-toggle__controls"
                    >
                      <span
                        class="pf-v5-c-menu-toggle__toggle-icon"
                      >
                        <svg
                          aria-hidden="true"
                          class="pf-v5-svg"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          viewBox="0 0 320 512"
                          width="1em"
                        >
                          <path
                            d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                          />
                        </svg>
                      </span>
                    </span>
                  </button>
                </div>
                <div
                  class="pf-v5-c-toolbar__item"
                >
                  <button
                    aria-expanded="false"
                    class="pf-v5-c-menu-toggle"
                    data-ouia-component-id="OUIA-Generated-MenuToggle-2"
                    data-ouia-component-type="PF5/MenuToggle"
                    data-ouia-safe="true"
                    type="button"
                  >
                    <span
                      class="pf-v5-c-menu-toggle__icon"
                    >
                      <div
                        style="display: contents;"
                      >
                        <span
                          class="sk-resource-icon sk-resource-process"
                          role="process-resource-icon"
                        >
                          P
                        </span>
                      </div>
                    </span>
                    <span
                      class="pf-v5-c-menu-toggle__text"
                    >
                      All source processes
                    </span>
                    <span
                      class="pf-v5-c-menu-toggle__controls"
                    >
                      <span
                        class="pf-v5-c-menu-toggle__toggle-icon"
                      >
                        <svg
                          aria-hidden="true"
                          class="pf-v5-svg"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          viewBox="0 0 320 512"
                          width="1em"
                        >
                          <path
                            d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                          />
                        </svg>
                      </span>
                    </span>
                  </button>
                </div>
                <div
                  class="pf-v5-c-toolbar__item"
                >
                  <button
                    aria-expanded="false"
                    class="pf-v5-c-menu-toggle"
                    data-ouia-component-id="OUIA-Generated-MenuToggle-3"
                    data-ouia-component-type="PF5/MenuToggle"
                    data-ouia-safe="true"
                    type="button"
                  >
                    <span
                      class="pf-v5-c-menu-toggle__icon"
                    >
                      <div
                        style="display: contents;"
                      >
                        <span
                          class="sk-resource-icon sk-resource-site"
                          role="site-resource-icon"
                        >
                          S
                        </span>
                      </div>
                    </span>
                    <span
                      class="pf-v5-c-menu-toggle__text"
                    >
                      All connected sites
                    </span>
                    <span
                      class="pf-v5-c-menu-toggle__controls"
                    >
                      <span
                        class="pf-v5-c-menu-toggle__toggle-icon"
                      >
                        <svg
                          aria-hidden="true"
                          class="pf-v5-svg"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          viewBox="0 0 320 512"
                          width="1em"
                        >
                          <path
                            d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                          />
                        </svg>
                      </span>
                    </span>
                  </button>
                </div>
                <div
                  class="pf-v5-c-toolbar__item"
                >
                  <button
                    aria-expanded="false"
                    class="pf-v5-c-menu-toggle"
                    data-ouia-component-id="OUIA-Generated-MenuToggle-4"
                    data-ouia-component-type="PF5/MenuToggle"
                    data-ouia-safe="true"
                    type="button"
                  >
                    <span
                      class="pf-v5-c-menu-toggle__icon"
                    >
                      <div
                        style="display: contents;"
                      >
                        <span
                          class="sk-resource-icon sk-resource-process"
                          role="process-resource-icon"
                        >
                          P
                        </span>
                      </div>
                    </span>
                    <span
                      class="pf-v5-c-menu-toggle__text"
                    >
                      All connected processes
                    </span>
                    <span
                      class="pf-v5-c-menu-toggle__controls"
                    >
                      <span
                        class="pf-v5-c-menu-toggle__toggle-icon"
                      >
                        <svg
                          aria-hidden="true"
                          class="pf-v5-svg"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          viewBox="0 0 320 512"
                          width="1em"
                        >
                          <path
                            d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                          />
                        </svg>
                      </span>
                    </span>
                  </button>
                </div>
                <div
                  class="pf-v5-c-toolbar__item"
                >
                  <button
                    aria-expanded="false"
                    class="pf-v5-c-menu-toggle"
                    data-ouia-component-id="OUIA-Generated-MenuToggle-5"
                    data-ouia-component-type="PF5/MenuToggle"
                    data-ouia-safe="true"
                    type="button"
                  >
                    <span
                      class="pf-v5-c-menu-toggle__text"
                    >
                      All protocols
                    </span>
                    <span
                      class="pf-v5-c-menu-toggle__controls"
                    >
                      <span
                        class="pf-v5-c-menu-toggle__toggle-icon"
                      >
                        <svg
                          aria-hidden="true"
                          class="pf-v5-svg"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          viewBox="0 0 320 512"
                          width="1em"
                        >
                          <path
                            d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                          />
                        </svg>
                      </span>
                    </span>
                  </button>
                </div>
              </div>
              <div
                class="pf-v5-c-toolbar__group pf-m-align-right"
              >
                <div
                  class="pf-v5-c-toolbar__item"
                >
                  <button
                    aria-expanded="false"
                    class="pf-v5-c-menu-toggle"
                    data-ouia-component-id="OUIA-Generated-MenuToggle-6"
                    data-ouia-component-type="PF5/MenuToggle"
                    data-ouia-safe="true"
                    data-testid="sk-time-range-filter-type"
                    type="button"
                  >
                    <span
                      class="pf-v5-c-menu-toggle__text"
                    >
                      <svg
                        aria-hidden="true"
                        class="pf-v5-svg"
                        fill="currentColor"
                        height="1em"
                        role="img"
                        viewBox="0 0 512 512"
                        width="1em"
                      >
                        <path
                          d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm0 448c-110.5 0-200-89.5-200-200S145.5 56 256 56s200 89.5 200 200-89.5 200-200 200zm61.8-104.4l-84.9-61.7c-3.1-2.3-4.9-5.9-4.9-9.7V116c0-6.6 5.4-12 12-12h32c6.6 0 12 5.4 12 12v141.7l66.8 48.6c5.4 3.9 6.5 11.4 2.6 16.8L334.6 349c-3.9 5.3-11.4 6.5-16.8 2.6z"
                        />
                      </svg>
                       Last 1 min.
                    </span>
                    <span
                      class="pf-v5-c-menu-toggle__controls"
                    >
                      <span
                        class="pf-v5-c-menu-toggle__toggle-icon"
                      >
                        <svg
                          aria-hidden="true"
                          class="pf-v5-svg"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          viewBox="0 0 320 512"
                          width="1em"
                        >
                          <path
                            d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                          />
                        </svg>
                      </span>
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div
            class="pf-v5-c-toolbar__content pf-m-hidden"
            hidden=""
          >
            <div
              class="pf-v5-c-toolbar__group"
            />
          </div>
        </div>
      </section>
    </div>
    <div
      class="pf-v5-l-stack__item"
    >
      <div
        class="pf-v5-c-card pf-m-expanded"
        data-ouia-component-id="OUIA-Generated-Card-1"
        data-ouia-component-type="PF5/Card"
        data-ouia-safe="true"
        id=""
      >
        <div
          class="pf-v5-c-card__header"
        >
          <div
            class="pf-v5-c-card__header-toggle"
          >
            <button
              aria-disabled="false"
              class="pf-v5-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-1"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe="true"
              type="button"
            >
              <span
                class="pf-v5-c-card__header-toggle-icon"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            class="pf-v5-c-card__header-main"
          >
            <div
              class="pf-v5-c-card__title"
            >
              <div
                class="pf-v5-c-card__title-text"
              >
                Traffic
              </div>
            </div>
          </div>
        </div>
        <div
          class="pf-v5-c-card__expandable-content"
        >
          <div
            class="pf-v5-c-card__body"
            style="min-height: 450px;"
          >
            <div
              style="position: absolute; width: 100%; height: 100%; background-color: rgba(3, 3, 3, 0.1); left: 0px; right: 0px; top: 0px; bottom: 0px; z-index: 1;"
            >
              <div
                class="pf-v5-l-bullseye"
                data-testid="sk-loading-view"
              >
                <svg
                  aria-label="Contents"
                  aria-valuetext="Loading..."
                  class="pf-v5-c-spinner pf-m-xl"
                  role="progressbar"
                  style="--pf-v5-c-spinner--diameter: 150px;"
                  viewBox="0 0 100 100"
                >
                  <circle
                    class="pf-v5-c-spinner__path"
                    cx="50"
                    cy="50"
                    fill="none"
                    r="45"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="pf-v5-l-stack__item"
    >
      <div
        class="pf-v5-c-card"
        data-ouia-component-id="OUIA-Generated-Card-2"
        data-ouia-component-type="PF5/Card"
        data-ouia-safe="true"
        id=""
      >
        <div
          class="pf-v5-c-card__header"
        >
          <div
            class="pf-v5-c-card__header-toggle"
          >
            <button
              aria-disabled="false"
              class="pf-v5-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-2"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe="true"
              type="button"
            >
              <span
                class="pf-v5-c-card__header-toggle-icon"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            class="pf-v5-c-card__header-main"
          >
            <div
              class="pf-v5-c-card__title"
            >
              <div
                class="pf-v5-c-card__title-text"
              >
                Tcp connections
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="pf-v5-l-stack__item"
    >
      <div
        class="pf-v5-c-card"
        data-ouia-component-id="OUIA-Generated-Card-3"
        data-ouia-component-type="PF5/Card"
        data-ouia-safe="true"
        id=""
      >
        <div
          class="pf-v5-c-card__header"
        >
          <div
            class="pf-v5-c-card__header-toggle"
          >
            <button
              aria-disabled="false"
              class="pf-v5-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-3"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe="true"
              type="button"
            >
              <span
                class="pf-v5-c-card__header-toggle-icon"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            class="pf-v5-c-card__header-main"
          >
            <div
              class="pf-v5-c-card__title"
            >
              <div
                class="pf-v5-c-card__title-text"
              >
                Incoming Latency 
                <div
                  style="display: contents;"
                >
                  <button
                    aria-disabled="false"
                    class="pf-v5-c-button pf-m-plain"
                    data-ouia-component-id="OUIA-Generated-Button-plain-4"
                    data-ouia-component-type="PF5/Button"
                    data-ouia-safe="true"
                    type="button"
                  >
                    <span
                      class="pf-v5-c-icon"
                    >
                      <span
                        class="pf-v5-c-icon__content pf-m-info"
                      >
                        <svg
                          aria-hidden="true"
                          class="pf-v5-svg"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          viewBox="0 0 512 512"
                          width="1em"
                        >
                          <path
                            d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zM262.655 90c-54.497 0-89.255 22.957-116.549 63.758-3.536 5.286-2.353 12.415 2.715 16.258l34.699 26.31c5.205 3.947 12.621 3.008 16.665-2.122 17.864-22.658 30.113-35.797 57.303-35.797 20.429 0 45.698 13.148 45.698 32.958 0 14.976-12.363 22.667-32.534 33.976C247.128 238.528 216 254.941 216 296v4c0 6.627 5.373 12 12 12h56c6.627 0 12-5.373 12-12v-1.333c0-28.462 83.186-29.647 83.186-106.667 0-58.002-60.165-102-116.531-102zM256 338c-25.365 0-46 20.635-46 46 0 25.364 20.635 46 46 46s46-20.636 46-46c0-25.365-20.635-46-46-46z"
                          />
                        </svg>
                      </span>
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="pf-v5-l-stack__item"
    >
      <div
        class="pf-v5-c-card"
        data-ouia-component-id="OUIA-Generated-Card-4"
        data-ouia-component-type="PF5/Card"
        data-ouia-safe="true"
        id=""
      >
        <div
          class="pf-v5-c-card__header"
        >
          <div
            class="pf-v5-c-card__header-toggle"
          >
            <button
              aria-disabled="false"
              class="pf-v5-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-5"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe="true"
              type="button"
            >
              <span
                class="pf-v5-c-card__header-toggle-icon"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            class="pf-v5-c-card__header-main"
          >
            <div
              class="pf-v5-c-card__title"
            >
              <div
                class="pf-v5-c-card__title-text"
              >
                Outgoing Latency 
                <div
                  style="display: contents;"
                >
                  <button
                    aria-disabled="false"
                    class="pf-v5-c-button pf-m-plain"
                    data-ouia-component-id="OUIA-Generated-Button-plain-6"
                    data-ouia-component-type="PF5/Button"
                    data-ouia-safe="true"
                    type="button"
                  >
                    <span
                      class="pf-v5-c-icon"
                    >
                      <span
                        class="pf-v5-c-icon__content pf-m-info"
                      >
                        <svg
                          aria-hidden="true"
                          class="pf-v5-svg"
                          fill="currentColor"
                          height="1em"
                          role="img"
                          viewBox="0 0 512 512"
                          width="1em"
                        >
                          <path
                            d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zM262.655 90c-54.497 0-89.255 22.957-116.549 63.758-3.536 5.286-2.353 12.415 2.715 16.258l34.699 26.31c5.205 3.947 12.621 3.008 16.665-2.122 17.864-22.658 30.113-35.797 57.303-35.797 20.429 0 45.698 13.148 45.698 32.958 0 14.976-12.363 22.667-32.534 33.976C247.128 238.528 216 254.941 216 296v4c0 6.627 5.373 12 12 12h56c6.627 0 12-5.373 12-12v-1.333c0-28.462 83.186-29.647 83.186-106.667 0-58.002-60.165-102-116.531-102zM256 338c-25.365 0-46 20.635-46 46 0 25.364 20.635 46 46 46s46-20.636 46-46c0-25.365-20.635-46-46-46z"
                          />
                        </svg>
                      </span>
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="pf-v5-l-stack__item"
    >
      <div
        class="pf-v5-c-card"
        data-ouia-component-id="OUIA-Generated-Card-5"
        data-ouia-component-type="PF5/Card"
        data-ouia-safe="true"
        id=""
      >
        <div
          class="pf-v5-c-card__header"
        >
          <div
            class="pf-v5-c-card__header-toggle"
          >
            <button
              aria-disabled="false"
              class="pf-v5-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-7"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe="true"
              type="button"
            >
              <span
                class="pf-v5-c-card__header-toggle-icon"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            class="pf-v5-c-card__header-main"
          >
            <div
              class="pf-v5-c-card__title"
            >
              <div
                class="pf-v5-c-card__title-text"
              >
                Http requests
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="pf-v5-l-stack__item"
    >
      <div
        class="pf-v5-c-card"
        data-ouia-component-id="OUIA-Generated-Card-6"
        data-ouia-component-type="PF5/Card"
        data-ouia-safe="true"
        id=""
      >
        <div
          class="pf-v5-c-card__header"
        >
          <div
            class="pf-v5-c-card__header-toggle"
          >
            <button
              aria-disabled="false"
              class="pf-v5-c-button pf-m-plain"
              data-ouia-component-id="OUIA-Generated-Button-plain-8"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe="true"
              type="button"
            >
              <span
                class="pf-v5-c-card__header-toggle-icon"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 256 512"
                  width="1em"
                >
                  <path
                    d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                  />
                </svg>
              </span>
            </button>
          </div>
          <div
            class="pf-v5-c-card__header-main"
          >
            <div
              class="pf-v5-c-card__title"
            >
              <div
                class="pf-v5-c-card__title-text"
              >
                Http responses
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
